Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,509 @@
+        //
+        // I think we should consider handling that as a fatalError, as fundamentally the ring size is too small
+        // compared to the amount of events the user tries to push through in a single eventloop tick.
+        //
+        // This is mostly a problem for synthetic tests that e.g. do a huge amount of registration modifications.
+        //
+        // This is a slight design issue with SwiftNIO in general that should be discussed.
+        //
+        while true {
+            if let sqe = CNIOLinux.io_uring_get_sqe(&ring) {
+                return sqe
+            }
+            self.io_uring_flush()
+            // FIXME(hassila): As we've added basic support for SQPOLL, we should use io_uring_sqring_wait() for waiting too
+        }
+    }
+
+    // Ok, this was a bummer - turns out that flushing multiple SQE:s
+    // can fail midflight and this will actually happen for real when e.g. a socket
+    // has gone down and we are re-registering polls this means we will silently lose any
+    // entries after the failed fd. Ouch. Proper approach is to use io_uring_sq_ready() in a loop.
+    // See: https://github.com/axboe/liburing/issues/309
+    internal func io_uring_flush() {         // When using SQPOLL this is basically a NOP
+        var waitingSubmissions : UInt32 = 0
+        var submissionCount = 0
+        var retval : Int32
+
+        waitingSubmissions = CNIOLinux.io_uring_sq_ready(&ring)
+
+        loop: while (waitingSubmissions > 0)
+        {
+            retval = CNIOLinux.io_uring_submit(&ring)
+            submissionCount += 1
+
+            switch retval {
+            case -EBUSY:
+                fallthrough
+            // We can get -EAGAIN if the CQE queue is full and we get back pressure from
+            // the kernel to start processing CQE:s. If we break here with unsubmitted
+            // SQE:s, they will stay pending on the user-level side and be flushed
+            // to the kernel after we had the opportunity to reap more CQE:s
+            // In practice it will be at the end of whenReady the next
+            // time around. Given the async nature, this is fine, we will not
+            // lose any submissions. We could possibly still get stuck
+            // trying to get new SQE if the actual SQE queue is full, but
+            // that would be due to user error in usage IMHO and we should fatalError there.
+            case -EAGAIN:
+                _debugPrint("io_uring_flush io_uring_submit -EBUSY/-EAGAIN waitingSubmissions[\(waitingSubmissions)] submissionCount[\(submissionCount)]. Breaking out and resubmitting later (whenReady() end).")
+                break loop
+            // -ENOMEM when there is not enough memory to do internal allocations on the kernel side.
+            // Right nog we just loop with a sleep trying to buy time, but could also possibly fatalError here.
+            // See: https://github.com/axboe/liburing/issues/309
```

Comment:
isn't https://github.com/axboe/liburing/issues/310 supposed to fix this? What's the state there?

Reply:
No, https://github.com/axboe/liburing/issues/310 is about being able to do updates to existing polls (especially multishot polls), the kernel could still presumably fail with internal memory allocations (however unlikely, it was just pointed out by one of the kernel developers as a possible failure state when [discussing the appropriate approach](https://github.com/axboe/liburing/issues/309).